ocrd/ocrd/mets_server.py

```diff
@@ -0,0 +1,237 @@
+from requests_unixsocket import Session as requests_unixsocket_session
+from pydantic import BaseModel, Field, constr, ValidationError
+
+import uvicorn
+
+from ocrd_models import OcrdMets
+from ocrd_utils import initLogging, getLogger, deprecated_alias
+
+#
+# XXX HACKS TODO
+#
+initLogging()
+
+#
+# Models
+#
+
+class OcrdFileModel(BaseModel):
+    file_grp : str = Field()
+    file_id : str = Field()
+    mimetype : str = Field()
+    page_id : Union[str, None] = Field()
+    url : str = Field()
+
+class OcrdAgentModel(BaseModel):
+    name : str = Field()
+    _type : str = Field()
+    role : str = Field()
+    otherrole : str = Field()
+    othertype : str = Field()
+    notes : List[Tuple[Dict[str, str], Optional[str]]] = Field()
+
+class OcrdFileListModel(BaseModel):
+    files : List[OcrdFileModel] = Field()
+
+class OcrdFileGroupListModel(BaseModel):
+    file_groups : List[str] = Field()
+
+#
+# Client
+#
+
+class ClientSideOcrdFile:
+    """
+    Provides the same interface as :py:class:`ocrd_models.ocrd_file.OcrdFile`
+    but without attachment to :py:class:`ocrd_models.ocrd_mets.OcrdMets` since
+    this represents the response of the :py:class:`ocrd.mets_server.OcrdWorkspaceServer`.
+    """
+
+    def __init__(self, el, mimetype=None, pageId=None, loctype='OTHER', local_filename=None, mets=None, url=None, ID=None, fileGrp=None):
+        """
+        Args:
+            el (): ignored
+        Keyword Args:
+            mets (): ignored
+            mimetype (string): ``@MIMETYPE`` of this ``mets:file``
+            pageId (string): ``@ID`` of the physical ``mets:structMap`` entry corresponding to this ``mets:file``
+            loctype (string): ``@LOCTYPE`` of this ``mets:file``
+            local_filename (): ignored
+            url (string): ``@xlink:href`` of this ``mets:file``
+            ID (string): ``@ID`` of this ``mets:file``
+        """
+        self.ID = ID
+        self.mimetype = mimetype
+        self.url = url
+        self.loctype = loctype
+        self.pageId = pageId
+        self.fileGrp = fileGrp
+
+class ClientSideOcrdMets():
+    """
+    Replacement for :py:class:`ocrd_models.ocrd_mets.OcrdMets` with overrides for
+    :py:meth:`ocrd_models.ocrd_mets.OcrdMets.find_files`,
+    :py:meth:`ocrd_models.ocrd_mets.OcrdMets.find_all_files`, and
+    :py:meth:`ocrd_models.ocrd_mets.OcrdMets.add_file` to query via HTTP a
+    :py:class:`ocrd.mets_server.OcrdMetsServer`.
+    """
+
+    def __init__(self, host, port, socket):
+        self.log = getLogger('ocrd.workspace_client')
+        if socket:
+            self.url = f'http+unix://{socket.replace("/", "%2F")}'
+            self.session = requests_unixsocket_session()
+        else:
+            self.url = f'http://{host}:{port}'
+            self.session = requests_session()
+
+    def find_files(self, **kwargs):
+        r = self.session.request('GET', self.url, params={**kwargs})
+        for f in r.json()['files']:
```

Comment:
Good idea in principle, it's just a bit tricky to test this. But since the METS Server is supposed to be transparent, there is arguably never a need to talk to the METS server via `curl` or `httpie`. Except to maybe tell it to sync to disk or stop the server, which don't require a request or response body. I'll look into it.

Reply:
In the interest of moving ahead, for now I don't want to add protobuf functionality unless there will be measurable performance issues.